File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/vab/provider/VABModelProvider.java
package org.eclipse.basyx.vab.provider;

import java.util.function.Function;

import org.eclipse.basyx.vab.core.IModelProvider;
import org.eclipse.basyx.vab.core.tools.VABPathTools;

/**
 * A generic VAB model provider.
 * 
 * @author espen
 */
public class VABModelProvider implements IModelProvider {
	/**
	 * Handler, which handles single element objects
	 */
	private IVABElementHandler handler;

	/**
	 * Root object that stores contained elements
	 */
	protected Object elements;

	public VABModelProvider(Object elements, IVABElementHandler handler) {
		this.handler = handler;
		this.elements = handler.preprocessObject(elements);
	}

	@Override
	public Object getModelPropertyValue(String path) throws Exception {
		// Check empty paths
		if (path == null) {
			throw new RuntimeException("Path is undefined");
		} else if (VABPathTools.isEmptyPath(path)) {
			return handler.postprocessObject(elements);
		}

		Object element = getTargetElement(path);
		return handler.postprocessObject(element);
	}

	@Override
	public void setModelPropertyValue(String path, Object newValue) throws Exception {
		// Check empty paths
		if (path == null) {
			throw new RuntimeException("Path is undefined");
		} else if (VABPathTools.isEmptyPath(path)) {
			// Empty path => parent element == null => replace root, if it exists
			if (elements != null) {
				elements = handler.preprocessObject(newValue);
			}
			return;
		}

		Object parentElement = getParentElement(path);
		String propertyName = VABPathTools.getLastElement(path);

		// Only write values, that already exist
		Object thisElement = handler.getElementProperty(parentElement, propertyName);
		if (parentElement != null && propertyName != null && thisElement != null) {
			newValue = handler.preprocessObject(newValue);
			handler.setModelPropertyValue(parentElement, propertyName, newValue);
		}
	}

	@Override
	public void createValue(String path, Object newValue) throws Exception {
		// Check empty paths
		if (path == null) {
			throw new RuntimeException("Path is undefined");
		} else if (VABPathTools.isEmptyPath(path)) {
			// The complete model should be replaced if it does not exist
			if (elements == null) {
				elements = handler.preprocessObject(newValue);
			}
			return;
		}

		// Find parent & name of new element
		Object parentElement = getParentElement(path);
		String propertyName = VABPathTools.getLastElement(path);

		// Only create new, never replace existing elements
		if (parentElement != null) {
			newValue = handler.preprocessObject(newValue);
			Object childElement = handler.getElementProperty(parentElement, propertyName);
			if (childElement == null) {
				handler.setModelPropertyValue(parentElement, propertyName, newValue);
			} else {
				handler.createValue(childElement, newValue);
			}
			return;
		}

		System.out.println("Could not create element, parent element does not exist for path '" + path + "'");
	}

	@Override
	public void deleteValue(String path) throws Exception {
		// Check null path
		if (path == null) {
			throw new RuntimeException("Path is undefined");
		}

		Object parentElement = getParentElement(path);
		String propertyName = VABPathTools.getLastElement(path);
		if (parentElement != null && propertyName != null) {
			handler.deleteValue(parentElement, propertyName);
		}
	}

	@Override
	public void deleteValue(String path, Object obj) throws Exception {
		// Check null path
		if (path == null) {
			throw new RuntimeException("Path is undefined");
		}

		Object parentElement = getParentElement(path);
		String propertyName = VABPathTools.getLastElement(path);
		if (parentElement != null && propertyName != null) {
			Object childElement = handler.getElementProperty(parentElement, propertyName);
			if (childElement != null) {
				handler.deleteValue(childElement, obj);
			}
		}
	}

	@SuppressWarnings("unchecked")
	@Override
	public Object invokeOperation(String path, Object[] parameters) throws Exception {
		Object childElement = getModelPropertyValue(path);

		// Invoke operation for function interfaces
		if (childElement != null && childElement instanceof Function<?, ?>) {
			Function<Object, Object[]> function = (Function<Object, Object[]>) childElement;
			return function.apply(parameters);
		}

		// No operation found
		return null;
	}

	/**
	 * Get the parent of an element in this provider. The path should include the path to the element separated by '/'.
	 * E.g., for accessing element c in path a/b, the path should be a/b/c.
	 */
	private Object getParentElement(String path) throws Exception {
		// Split path into its elements, separated by '/'
		String[] pathElements = VABPathTools.splitPath(path);

		Object currentElement = elements;
		// ignore the leaf element, only return the leaf's parent element
		for (int i = 0; i < pathElements.length - 1; i++) {
			if (currentElement == null) {
				return null;
			}
			currentElement = handler.getElementProperty(currentElement, pathElements[i]);
		}
		return currentElement;
	}

	/**
	 * Instead of returning the parent element of a path, this function gives the target element.
	 * E.g., it returns c for the path a/b/c
	 */
	protected Object getTargetElement(String path) throws Exception {
		if (VABPathTools.isEmptyPath(path)) {
			return elements;
		} else {
			Object parentElement = getParentElement(path);
			String operationName = VABPathTools.getLastElement(path);
			if (parentElement != null && operationName != null) {
				return handler.getElementProperty(parentElement, operationName);
			}
		}
		return null;
	}
}

File: sdks/c++/basys.sdk.cc/src/vab/vab/provider/VABModelProvider.cpp
#include <BaSyx/vab/provider/VABModelProvider.h>


using namespace basyx::vab::provider;

VABModelProvider::VABModelProvider(const basyx::object& any) 
	: elements(any) 
{
};

basyx::object VABModelProvider::getParentElement(const std::string& path)
{
	core::VABPath vabPath{ path };

	auto currentElement = elements;

	// ignore the leaf element, only return the leaf's parent element
	const auto & pathElements = vabPath.getElements();
	for(std::size_t i=0; i < pathElements.size()-1; ++i)
	{
		if (currentElement.IsNull())
			return basyx::object::make_null();

		currentElement = currentElement.getProperty(pathElements[i]); //  handler.GetElementProperty(currentElement, pathElements[i]);
	}

	return currentElement;
};

basyx::object VABModelProvider::getTargetElement(const std::string& path)
{
	core::VABPath vabPath{ path };
	if (vabPath.isEmpty()) {
		return elements;
	}
	else
	{
		auto parentElement = getParentElement(path);
		auto operationName = vabPath.getLastElement();
		if (!parentElement.IsNull() && !operationName.empty()) {
			//return handler.GetElementProperty(parentElement, operationName);
			return parentElement.getProperty(operationName);
		}
	};

	return basyx::object::make_null();
};

basyx::object VABModelProvider::getModelPropertyValue(const std::string& path)
{
	core::VABPath vabPath{ path };
	auto & lastPathElement = vabPath.getLastElement();

	if (vabPath.isEmpty()) {
		//return handler.PostProcessObject(elements);
		return elements;
	}

	auto element = getTargetElement(path);
//			return handler.PostProcessObject(element);
	return element;
}

void VABModelProvider::setModelPropertyValue(const std::string& path, const basyx::object newValue)
{
	core::VABPath vabPath{ path };

	// Check empty paths
	if (vabPath.isEmpty()) {
		// Empty path => parent element == null => replace root, if it exists
		if (!elements.IsNull()) {
			elements = newValue;
		}
		return;
	}

	auto parentElement = this->getParentElement(path);
	auto & propertyName = vabPath.getLastElement();

	// Only write values, that already exist
	auto thisElement = parentElement.getProperty(propertyName);

	if (!parentElement.IsNull() && !thisElement.IsNull()) {
		parentElement.insertKey(propertyName, newValue, true);
	}
};

void VABModelProvider::createValue(const std::string& path, const basyx::object newValue)
{
	core::VABPath vabPath{ path };

	// Check empty paths
	if (vabPath.isEmpty()) {
		// The complete model should be replaced if it does not exist
		if (elements.IsNull()) {
			elements = newValue;
		}
		return;
	}

	// Find parent & name of new element
	auto parentElement = this->getParentElement(path);
	auto & propertyName = vabPath.getLastElement();

	// Only create new, never replace existing elements
	if(!parentElement.IsNull()) {
		auto childElement = parentElement.getProperty(propertyName);
		if (childElement.IsNull()) {
			parentElement.insertKey(propertyName, newValue);
		}
		else {
			childElement.insert(newValue);
		}
		return;
	}

	log.warn("Could not create element, parent element does not exist for path '{}'", path);
};

void VABModelProvider::deleteValue(const std::string& path, basyx::object deletedValue)
{
	core::VABPath vabPath{ path };

	if (vabPath.isEmpty())
		return;

	// Find parent & name of element
	auto parentElement = this->getParentElement(path);
	auto & propertyName = vabPath.getLastElement();

	if (!parentElement.IsNull()) {
		//auto & childElement = handler.GetElementProperty(parentElement, propertyName);
		auto childElement = parentElement.getProperty(propertyName);


		if (!childElement.IsNull()) {
			//handler.DeleteValue(childElement, deletedValue);
			childElement.remove(deletedValue);
		}
	}
};

void VABModelProvider::deleteValue(const std::string& path)
{
	core::VABPath vabPath{ path };

	// Find parent & name of element
	auto parentElement = this->getParentElement(path);
	auto & propertyName = vabPath.getLastElement();

	if (!parentElement.IsNull()) {
		//handler.DeleteProperty(parentElement, propertyName);
		parentElement.removeProperty(propertyName);
	}
};

basyx::object VABModelProvider::invokeOperation(const std::string& path, basyx::object parameters)
{
	log.trace("invokeOperationImpl called");
	log.trace("    path: \"{}\"", path);
	log.trace("    parameters: <not impl>");

	auto element = this->getModelPropertyValue(path);

	if (element.IsNull()) {
		log.error("Function not found!");
		log.trace("Returning basyx::object::null");

		return basyx::object{ nullptr };
	}

	if (!element.IsInvokable()) {
		log.error("Found object is not invokable!");
		log.trace("Returning basyx::object::null");

		return basyx::object{ nullptr };
	};

	log.trace("Function found. Invoking...");

	return element.invoke(parameters);
};